Add AI-Enhanced Documentation Generator (doc-gen) Best Practice

yunks128 opened the pull request:
**Purpose:**
This PR introduces a new best practice called `doc-gen`, which enables automatic generation of documentation sites using a Docusaurus-based template. It supports AI enhancements using models like `ollama/gemma3` or `openai/gpt-4o-mini`.

**Proposed Changes:**

* \[ADD] `docgen.py`: New best practice module for generating and revising documentation sites.
* \[ADD] CLI support for `--output-dir`, `--template-only`, `--revise-site` flags.
* \[ADD] Auto-installation check for `slim-doc-generator`.
* \[ADD] Registry integration and fallback support for `doc-gen`.
* \[CHANGE] Updated supported models (`ollama/gemma3`, `openai/gpt-4o-mini`).
* \[CHANGE] Updated CLI logic to cleanly route arguments for `doc-gen`.
* \[CHANGE] Documentation (`README.md`) to show new usage examples.

**Issues:**
#11

**Testing:**
✅ Verified CLI command runs for:

- Template-Only Mode - Generate just the template structure without analyzing a repository:
   ```bash
   slim apply --best-practice-ids doc-gen --template-only --output-dir ../docs-site
   ```
   - AI-Enhanced Documentation
   ```bash
   slim apply --best-practice-ids doc-gen --repo-dir /path/to/your/repo --output-dir /path/to/output --use-ai "ollama/gemma3"
   ```
   Example usage:
   ```bash
   slim apply --best-practice-ids doc-gen --repo-dir ./hysds --output-dir ./hysds-docs-site --use-ai "openai/gpt-4o-mini"
   ```
   - Revising an Existing Site
   ```bash
   slim apply --best-practice-ids doc-gen --revise-site --output-dir ./docs-site --use-ai "openai/gpt-4o-mini"
   ```

riverma reviewed (changes requested):
Hey @yunks128 - great work here moving this forward! I think it'll be best to address some changes / improvements before we make this customer facing. Some suggestions:
- Deprecate old documentation generator code / README references
- Ensure that, when generating docs via AI, the title of the created folder code's package.json doesn't contain illegal characters. I got this problem (described below) when trying with `gemma3`.
- The quality of the site using `gemma3` isn't that great. I think the README should explicitly recommend using cloud models, and mention using local models is not likely to give great results. The only other options are OpenAI. What if we use this PR to upgrade the AI code to be more extensible and support more models out-of-the-box? My suggestion is to use https://docs.litellm.ai/docs/ or LangChain. Thoughts?
- Is there a way we can get a progress bar as the doc site is being generated?

yunks128 commented:
@riverma Thanks for your feedback! I finally moved everything into slim-cli! Here are the changes! 

### Documentation Generation Engine
- **`src/jpl/slim/docgen/`** - Complete documentation generation system
  - `generator.py` - Main SlimDocGenerator class
  - `site_reviser.py` - Updates existing sites with AI
  - `analyzer/` - Repository analysis and content extraction
  - `content/` - Section generators (overview, installation, API, etc.)
  - `enhancer/` - AI content enhancement
  - `template/` - Docusaurus template management

### Best Practice Implementation  
- **`src/jpl/slim/best_practices/docgen.py`** - DocGenPractice class for doc-gen best practice

### New Commands
- **`src/jpl/slim/commands/models_command.py`** - AI model discovery and management
- **`src/jpl/slim/commands/examples.py`** - Usage examples and help text

### Tests
- **`tests/jpl/slim/test_litellm_integration.py`** - LiteLLM integration tests

## Major File Modifications

### CLI Core (`src/jpl/slim/cli.py`)
- Added LiteLLM integration and model validation
- Enhanced argument parsing and dry-run support  
- Added startup banner and better error handling

### AI Utils (`src/jpl/slim/utils/ai_utils.py`)
- Added `generate_with_litellm()` for 100+ AI models
- Added model recommendation system
- Added provider setup instructions

### Commands Common (`src/jpl/slim/commands/common.py`)
- Added model availability checking
- Added provider configuration helpers
- Enhanced AI model support infrastructure

### Apply Command (`src/jpl/slim/commands/apply_command.py`)
- Added doc-gen specific arguments (`--output-dir`, `--template-only`, `--revise-site`)
- Enhanced parameter passing to best practices

### Best Practices Manager (`src/jpl/slim/manager/best_practices_manager.py`)
- Added DocGenPractice support
- Added SLIM-17.1 alias mapping

## Configuration Updates

### Dependencies (`pyproject.toml`)
```toml
+ 'pyyaml ~= 6.0.1'
+ 'jinja2 ~= 3.1.0' 
+ 'click ~= 8.1.0'
+ 'litellm>=1.40.0'
```

### Package Exports (`src/jpl/slim/best_practices/__init__.py`)
```python
+ from jpl.slim.best_practices.docgen import DocGenPractice
+ __all__ = [..., "DocGenPractice"]
```

## Removed Files
- **`src/jpl/slim/commands/generate_docs_command.py`** - Replaced by integrated doc-gen best practice

## Key Usage Changes

### New Documentation Generation
```bash
slim apply --best-practice-ids doc-gen --repo-dir ./project --output-dir ./docs --use-ai openai/gpt-4o
```

### New Model Management  
```bash
slim models list                    # Discover available AI models
slim models recommend --task documentation   # Get recommendations  
slim models setup anthropic         # Setup instructions
```

riverma commented:
Awesome @yunks128! That's a massive step, thanks for taking the time! Let me give this a whirl  on my laptop and get back to you on the PR. Thank you again for the hard work!!

riverma reviewed (changes requested):
Hey @yunks128 - amazing job here! I really like the usability improvements and the port over from the old code, plus new litellm support for so many models! Incredible job.

Some observations:
1. Unit tests do not all pass via `pytest`. We'll want those all to pass.
2. I'm still getting errors after the doc-gen is complete (with AI support), such as the below, where the title contains a ":". Can we prevent this somehow?
    ```
      [cause]: YAMLException: incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line at line 4, column 12:
      title: MMTC: Multi-Mission Time Correlation ... 
    ```
3. "LiteLLM is not available" warning appears in Python 3.13 because litellm has a dependency on the imghdr module, which was removed in Python 3.13. This causes the import to fail even though litellm is installed via pip. The issue occurs when the CLI tries to import litellm at startup in cli.py, catches the ImportError, and displays the warning message. To resolve this, users would need to either:
  - Use Python 3.12 or earlier (as imghdr was removed in 3.13)
  - Wait for litellm to update their package for Python 3.13 compatibility
  - Consider making the litellm import lazy to avoid the warning on startup for users who don't need AI features
4. SonarQube has flagged some issues

## UPDATE June 5th, 2025

@yunks128 - I've made a PR that resolves the first two. See #36

riverma commented:
Some updates:
- `pytest` now passes
- No more `YAMLException` with generated site

To do's:
- [ ] - Model providers like Google don't seem to be aligned with the latest documentation for LiteLLM. I get model not found. Make sure that all the models listed are accurate in terms of what is actually provided by LiteLLM
- [ ] - The code is currently using an embedded `slim-registry.json`. We should remove this and only use the remote `slim-registry.json` file
- [ ] - Usability improvements
